### PR TITLE
ci: enrich validation step to take all the posible reana.yaml files

### PR DESCRIPTION
**Pull request by marcdiazsan**
Upgrade python2.7 to python3.8

Closes reanahub/reana-demo-helloworld#54

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -18,15 +18,18 @@ jobs:
...
-          pip install yadage
+          pip install --pre reana-client
 
-      - name: Validate workflow
-        run: yadage-validate workflow/yadage/workflow.yaml | grep -q 'workflow validates'
+      - name: Validate workflows
+        run: |        
+          for file in $FILES; do reana-client validate -f $file; done
+        env:
+          FILES: reana.yaml reana-slurmcern.yaml reana-htcondorcern.yaml reana-yadage.yaml reana-yadage-slurmcern.yaml reana-yadage-htcondorcern.yaml reana-cwl.yaml reana-cwl-slurmcern.yaml reana-cwl-htcondorcern.yaml
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:32]: `reana-snakemake.yaml?`

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -18,15 +18,18 @@ jobs:
...
 
-      - name: Validate workflow
-        run: yadage-validate workflow/yadage/workflow.yaml | grep -q 'workflow validates'
+      - name: Validate workflows
+        run: |        
+          for file in $FILES; do reana-client validate -f $file; done
+        env:
+          FILES: reana.yaml reana-slurmcern.yaml reana-htcondorcern.yaml reana-yadage.yaml reana-yadage-slurmcern.yaml reana-yadage-htcondorcern.yaml reana-cwl.yaml reana-cwl-slurmcern.yaml reana-cwl-htcondorcern.yaml 
 
       - run: rm -rf results && mkdir results
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:35]: Should we add this and the two `run` below to a single step `Run workflows`?
And we can also add sections for other workflow engines as @tiborsimko mentioned here:

> If we want to get extra fancy, we could introduce separate "Run workflow" tests with running on naked CWL/Yadage/Serial-on-Docker/Snakemake environment, using the usual out-of-REANA instructions that we document there, i.e. mkdir snakemake-local-run etc.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -18,15 +18,18 @@ jobs:
       - name: Setup Python
         uses: actions/setup-python@v2
         with:
-          python-version: 2.7
+          python-version: 3.8
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:21]: Nice! I guess we have this in other examples?

Comment (marcdiazsan) [reply .github/workflows/ci.yml:21]: Yes, we have `python2.7` in several demos such as `world_population` and `root6-roofit`. We also need to change the validation in some of them take off the `| grep -q 'workflow validates' ` and to add all the ´.yaml´ files in other ones. It's a little sprint to update them all. I'll make the corresponding PRs set to update all the demos

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -18,18 +18,20 @@ jobs:
...
-      - run: grep -q 'Hello Joe Bloggs!' results/greetings.txt
+          pip install --pre reana-client
+
+      - name: Validate workflows
+        run: |        
+          for file in $FILES; do reana-client validate -f $file; done
+        env:
+          FILES: reana.yaml reana-slurmcern.yaml reana-htcondorcern.yaml reana-yadage.yaml reana-yadage-slurmcern.yaml reana-yadage-htcondorcern.yaml reana-cwl.yaml reana-cwl-slurmcern.yaml reana-cwl-htcondorcern.yaml reana-snakemake.yaml
+
+      - run: |
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:34]: We can name it "Run workflows" maybe?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -18,18 +18,21 @@ jobs:
...
+          python-version: 3.8
 
       - name: Install Python dependencies
         run: |
           pip install --upgrade pip
-          pip install yadage
+          pip install --pre reana-client
 
-      - name: Validate workflow
-        run: yadage-validate workflow/yadage/workflow.yaml | grep -q 'workflow validates'
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:29]: I guess this part is covered by the [`yadage_load` function](https://github.com/reanahub/reana-commons/blob/294c0c59448e8d12c8772a29a5854838b08d4884/reana_commons/yadage.py#L16)?

Just trying to make sure if we can get rid of this safely.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -18,18 +18,21 @@ jobs:
...
-      - run: rm -rf results && mkdir results
-
-      - run: docker run -i --rm -v `pwd`/code:/code -v `pwd`/data:/data -v `pwd`/results:/results python:2.7-slim python /code/helloworld.py --sleeptime 0
-
-      - run: grep -q 'Hello Joe Bloggs!' results/greetings.txt
+      - name: Run workflows
+        run: |
+          rm -rf results && mkdir results
+          docker run -i --rm -v `pwd`/code:/code -v `pwd`/data:/data -v `pwd`/results:/results python:2.7-slim python /code/helloworld.py --sleeptime 0
+          grep -q 'Hello Joe Bloggs!' results/greetings.txt
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:38]: This covers serial workflows, should we add steps to test Yadage, CWL and Snakemake locally in a similar way?

> @tiborsimko: If we want to get extra fancy, we could introduce separate "Run workflow" tests with running on naked CWL/Yadage/Serial-on-Docker/Snakemake environment, using the usual out-of-REANA instructions that we document there, i.e. mkdir snakemake-local-run etc.

Comment (marcdiazsan) [reply .github/workflows/ci.yml:38]: Could you please remind me where are those instructions? I'm sure I have seen them but I cannot find them

File: .github/workflows/ci.yml (hunk 5 above)
Comment (marcdiazsan) [reply .github/workflows/ci.yml:29]: Yes I understand I covered by that function

File: .github/workflows/ci.yml (hunk 6 above)
Comment (mvidalgarcia) [reply .github/workflows/ci.yml:38]: - Yadage: https://github.com/reanahub/reana-demo-helloworld/blob/4599cdd365a59480abe6c1fc998c12f21f40ab5f/workflow/yadage/workflow.yaml#L1-L13
- CWL: https://github.com/reanahub/reana-demo-helloworld/blob/4599cdd365a59480abe6c1fc998c12f21f40ab5f/workflow/cwl/helloworld.cwl#L1-L12
- Snakemake: https://github.com/reanahub/reana-demo-helloworld/blob/4599cdd365a59480abe6c1fc998c12f21f40ab5f/workflow/snakemake/Snakefile#L1-L14

**Review by mvidalgarcia — COMMENTED** [review]
As we don't change the demos so often, it'd be nice to run these tests periodically to detect potential issues in latest versions.

GitHub Actions has a way to [schedule events](https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#scheduled-events) and run the tests periodically, similarly as a cronjob works. We can include this behavior in our demo tests to run them e.g. every Monday at 6am?

**Comment by marcdiazsan** [comment]
@mvidalgarcia I agree with the scheduled actions. I just wonder if that will trigger the actions at 7am your timezone or mine timezone

**Comment by mvidalgarcia** [comment]
> @mvidalgarcia I agree with the scheduled actions. I just wonder if that will trigger the actions at 7am your timezone or mine timezone

It's UTC, as stated on the docs.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -18,18 +23,60 @@ jobs:
...
+      - name: Validate workflows
+        run: |        
+          for file in $FILES; do reana-client validate --environments -f $file; done
+        env:
+          FILES: reana.yaml reana-slurmcern.yaml reana-htcondorcern.yaml reana-yadage.yaml reana-yadage-slurmcern.yaml reana-yadage-htcondorcern.yaml reana-cwl.yaml reana-cwl-slurmcern.yaml reana-cwl-htcondorcern.yaml reana-snakemake.yaml
 
-      - run: rm -rf results && mkdir results
-
-      - run: docker run -i --rm -v `pwd`/code:/code -v `pwd`/data:/data -v `pwd`/results:/results python:2.7-slim python /code/helloworld.py --sleeptime 0
+      - name: Run serial workflows
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:40]:
```suggestion
      - name: Run serial workflow
```

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -18,18 +23,60 @@ jobs:
...
+        run: |
+          rm -rf results && mkdir results
+          docker run -i --rm -v `pwd`/code:/code -v `pwd`/data:/data -v `pwd`/results:/results python:2.7-slim python /code/helloworld.py --sleeptime 0
+          grep -q 'Hello Joe Bloggs!' results/greetings.txt
+          rm -rf results
+      - name: Run yadage workflows
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:46]:
```suggestion
      - name: Run Yadage workflow
```

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -18,18 +23,60 @@ jobs:
...
+          cd yadage-local-run
+          virtualenv ~/.virtualenvs/yadage
+          source ~/.virtualenvs/yadage/bin/activate
+          pip install yadage
+          cp -a ../code ../data yadage-inputs
+          yadage-run . ../workflow/yadage/workflow.yaml -p sleeptime=0 -p inputfile=data/names.txt -p helloworld=code/helloworld.py -d initdir=`pwd`/yadage-inputs
+          grep -q 'Hello Joe Bloggs!' helloworld/greetings.txt
+          cd ..
+          rm -rf yadage-local-run
+      - name: Run cwl workflows
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:58]:
```suggestion
      - name: Run CWL workflow
```

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -18,18 +23,60 @@ jobs:
...
+          rm -rf yadage-local-run/yadage-inputs && mkdir -p yadage-local-run/yadage-inputs
+          cd yadage-local-run
+          virtualenv ~/.virtualenvs/yadage
+          source ~/.virtualenvs/yadage/bin/activate
+          pip install yadage
+          cp -a ../code ../data yadage-inputs
+          yadage-run . ../workflow/yadage/workflow.yaml -p sleeptime=0 -p inputfile=data/names.txt -p helloworld=code/helloworld.py -d initdir=`pwd`/yadage-inputs
+          grep -q 'Hello Joe Bloggs!' helloworld/greetings.txt
+          cd ..
+          rm -rf yadage-local-run
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:57]: You can also remove the virtualenv
```suggestion
          rm -rf yadage-local-run ~/.virtualenvs/yadage
```

Same for the other workflow languages.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -18,18 +23,60 @@ jobs:
...
+          cd cwl-local-run
+          virtualenv ~/.virtualenvs/cwl
+          source ~/.virtualenvs/cwl/bin/activate
+          pip install cwltool
+          cp -a ../code ../data . && cp ../workflow/cwl/helloworld-job.yml .
+          cwltool --quiet --outdir="./results" ../workflow/cwl/helloworld.cwl helloworld-job.yml
+          grep -q 'Hello Joe Bloggs!' results/greetings.txt
+          cd ..
+          rm -rf cwl-local-run
+      - name: Run snakemake workflows
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:70]:
```suggestion
      - name: Run Snakemake workflow
```

**Review by mvidalgarcia — APPROVED** [review]
Looking great! 💅

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -18,18 +23,60 @@ jobs:
...
+          rm -rf snakemake-local-run && mkdir -p snakemake-local-run
+          cd snakemake-local-run
+          virtualenv ~/.virtualenvs/snakemake
+          source ~/.virtualenvs/snakemake/bin/activate
+          pip install snakemake
+          cp -a ../code ../data .
+          snakemake -s ../workflow/snakemake/Snakefile --configfile ../workflow/snakemake/inputs.yaml -p --cores 1
+          grep -q 'Hello Joe Bloggs!' results/greetings.txt
+          cd ..
+          rm -rf snakemake-local-run ~/.virtualenvs/cwl
```
Comment (mvidalgarcia) [inline .github/workflows/ci.yml:81]: Whoops
```suggestion
          rm -rf snakemake-local-run ~/.virtualenvs/snakemake
```

**Pull request merged**